### PR TITLE
fix: stateless s3 storage for verdaccio/verdaccio/issues/1595

### PR DESCRIPTION
**Pull request by favoyang**
**Type:** bug
**Scope:** aws-s3-storage

The following has been addressed in the PR:

*  There is a related issue? #317
*  Unit or Functional tests are included in the PR? No change

**Description:**
The PR implemented solution B mentioned in https://github.com/verdaccio/monorepo/issues/317#issuecomment-558977885, to make s3 backend stateless. ~~**It's not finished yet, but most for demostrate the idea.**~~

- no long use the verdaccio database file (the `verdaccio-s3-db.json`) to store package list. Instead, use the s3 list action, get all sub-directories of the bucket as the package list. i.e. 
```
verdaccio/
  package-a/
  package-b/
  @some-scope/
    package-c
  verdaccio-s3-db.json

==> ['package-a', 'package-b', '@some-scope/package-c']
```
- `verdaccio-s3-db.json` is still useful to store generated secret.
- scoped package is supported by performing the s3 list action on each `@scope` folders, and appending result to the final package list.
- The get package list generated `1 + num_of_scoped_packages` s3 list actions.

### CODE REVIEW

**Comment by favoyang** [comment]
> Interesting concept, could you rebase from master? I'd like to give it a try locally.

Will try to rebase and notify you later.

**Comment by favoyang** [comment]
@juanpicado master merged and did some cleanups, it's ready to review. However I noticed the index.ts tests are still skipped. I only did some manual tests.

**Comment by juanpicado** [comment]
Thanks :) I will check this if I have some time after work, thanks for rebase it !

File: yarn.lock [hunk 1]
```diff
@@ -3182,6 +3182,21 @@ atob@^2.1.2:
   resolved "https://registry.verdaccio.org/atob/-/atob-2.1.2.tgz#6d9517eb9e030d2436666651e86bd9f6f13533c9"
   integrity sha512-Wm6ukoaOGJi/73p/cl2GvLjTI5JM1k/O14isD73YML8StrH/7/lRFgmg8nICZgD3bZZvjwCGxtMOD3wWNAu8cg==
 
+aws-sdk@^2.415.0:
```
Comment (STRML) [inline yarn.lock]: Can be deduped with below aws-sdk

File: plugins/aws-s3-storage/src/index.ts [hunk 2]
```diff
@@ -227,6 +195,68 @@ export default class S3Database implements IPluginStorage<S3Config> {
...
+            });
+          })
+        )
+          .then(() => tree.packages)
+          .then(packages => {
+            packages.forEach(pkg => {
+              logger.debug({ pkg }, 's3: [_getPackageList] package: @{pkg}');
+            });
+            return packages;
+          });
```
Comment (STRML) [inline plugins/aws-s3-storage/src/index.ts]:
```suggestion
          .then(() => {
            tree.packages.forEach(pkg => {
              logger.debug({ pkg }, 's3: [_getPackageList] package: @{pkg}');
            });
            return tree.packages;
          });
```

File: plugins/aws-s3-storage/src/index.ts [hunk 3]
```diff
@@ -227,6 +195,68 @@ export default class S3Database implements IPluginStorage<S3Config> {
...
+            tree.scopes.push(seg);
+          } else {
+            tree.packages.push(seg);
+          }
+        });
+        return Promise.resolve(tree);
+      });
+    }
+    return listS3Folder(listPath).then(tree => {
+      if (!tree.scopes.length) {
```
Comment (STRML) [inline plugins/aws-s3-storage/src/index.ts]: Branch is unneeded. If there are no scopes, the `Promise.all()` will resolve immediately and we'll simply return `tree.packages` as we expect.

File: plugins/aws-s3-storage/src/index.ts [hunk 4]
```diff
@@ -227,6 +195,68 @@ export default class S3Database implements IPluginStorage<S3Config> {
...
+        folders.forEach(folder => {
+          const segs = folder.split('/');
+          const seg = segs[segs.length - 2];
+          if (seg.startsWith('@')) {
+            tree.scopes.push(seg);
+          } else {
+            tree.packages.push(seg);
+          }
+        });
+        return Promise.resolve(tree);
```
Comment (STRML) [inline plugins/aws-s3-storage/src/index.ts]: Unneeded `Promise.resolve()`

File: plugins/aws-s3-storage/src/index.ts [hunk 5]
```diff
@@ -227,6 +195,68 @@ export default class S3Database implements IPluginStorage<S3Config> {
     return this._localData as LocalStorage;
   }
 
+  private async _getPackageList(): Promise<any> {
+    const logger = this.logger;
+    const { bucket, keyPrefix } = this.config;
+    const listPath = keyPrefix.endsWith('/') ? keyPrefix : '/';
```
Comment (STRML) [inline plugins/aws-s3-storage/src/index.ts]: This shouldn't need editing. See `addTrailingSlash` helper added in #313 (https://github.com/verdaccio/monorepo/pull/313/files#diff-362e190e13ca3d0e1e54d19c7a3f45e0R38)

File: yarn.lock (hunk 1 above)
Comment (favoyang) [reply yarn.lock]: Will remove

File: plugins/aws-s3-storage/src/index.ts (hunk 2 above)
Comment (favoyang) [reply plugins/aws-s3-storage/src/index.ts]: Looks good

File: plugins/aws-s3-storage/src/index.ts (hunk 3 above)
Comment (favoyang) [reply plugins/aws-s3-storage/src/index.ts]: Looks good

File: plugins/aws-s3-storage/src/index.ts (hunk 4 above)
Comment (favoyang) [reply plugins/aws-s3-storage/src/index.ts]: Looks good

File: plugins/aws-s3-storage/src/index.ts (hunk 5 above)
Comment (favoyang) [reply plugins/aws-s3-storage/src/index.ts]: Looks good

File: yarn.lock (hunk 1 above)
Comment (favoyang) [reply yarn.lock]: Well it's the library s3-ls@^3.0.0 depends on aws-sdk "^2.415.0". I guess I can not do much here?

**Comment by favoyang** [comment]
@STRML all code review fixed, except the `yarn.lock` dedup one.

File: yarn.lock (hunk 1 above)
Comment (STRML) [reply yarn.lock]: You can. Delete both aws-sdk entries from yarn.lock, then run `yarn` again. You should end up with an entry like:

```
aws-sdk@^2.415.0, aws-sdk@^2.607.0:
  version "2.612.0"
```

You can also use https://github.com/atlassian/yarn-deduplicate to help out with this.

Comment (favoyang) [reply yarn.lock]: Thanks for the tip. I cloud be be part of the yarn CLI. Anyway all CR are fixed.

**Comment by favoyang** [comment]
Worth reading: https://github.com/barolab/verdaccio-minio/issues/13

**Comment by benjamine** [comment]
This is so close :), is there anything beyond conflict solving we can help here? Thanks!

**Comment by favoyang** [comment]
@benjamine I replied to you on https://github.com/verdaccio/monorepo/issues/317#issuecomment-719641940, to keep the original discussion continued.

**Comment by favoyang** [comment]
The PR is abandoned due to https://github.com/verdaccio/monorepo/issues/317#issuecomment-719641940.

**Pull request closed without merging**